### PR TITLE
Small bootstrap improvements

### PR DESCRIPTION
**Pull request by kadiwa4**
- i/o-less check for `xz` availability
- cache result of NixOS detection
- load correct `bootstrap` module even when a package of that name is installed
- no `-W semicolon_in_expressions_from_macros` – it is warn-by-default
- one type per variable (making dynamic typing less confusing)
- integrate python-side `--help` flag into the argument parser (makes `-hv` work as a short form of `--help --verbose`)

I even checked that it works with Python 2.

### CODE REVIEW

File: src/bootstrap/download.rs [hunk 1]
```diff
@@ -70,53 +70,62 @@ impl Config {
...
-        // At this point we're pretty sure the user is running NixOS or using Nix
-        println!("info: you seem to be using Nix. Attempting to patch {}", fname.display());
+    /// Modifies the interpreter section of 'fname' to fix the dynamic linker,
+    /// or the RPATH section, to fix the dynamic library search path
+    ///
+    /// This is only required on NixOS and uses the PatchELF utility to
+    /// change the interpreter/RPATH of ELF executables.
+    ///
+    /// Please see https://nixos.org/patchelf.html for more information
+    fn fix_bin_or_dylib(&self, fname: &Path) {
```
Comment (albertlarsan68) [inline src/bootstrap/download.rs:127]: Can you add an assert making sure that this function is only called if `should_fix_bins_and_dylibs` returns `true` (and same for the python side of things)

**Comment by jyn514** [comment]
> cache result of NixOS detection

Could you please do that in the Rust code too?

File: src/bootstrap/bootstrap.py [hunk 2]
```diff
@@ -395,15 +394,15 @@ class RustBuild(object):
...
         self.build_dir = ''
         self.clean = False
         self.config_toml = ''
         self.rust_root = ''
-        self.use_locked_deps = ''
-        self.use_vendored_sources = ''
-        self.verbose = False
+        self.use_locked_deps = False
+        self.use_vendored_sources = False
+        self.verbose = 0
```
Comment (albertlarsan68) [inline src/bootstrap/bootstrap.py:405]: If the old variable could be kept, it would be better, since this allows keeping the simpler `self.verbose` arguments.
Maybe add `self.verbose_count`?

File: x.py [hunk 3]
```diff
@@ -22,7 +22,7 @@
             pass
 
 rust_dir = os.path.dirname(os.path.abspath(__file__))
-sys.path.append(os.path.join(rust_dir, "src", "bootstrap"))
+sys.path[:0] = [os.path.join(rust_dir, "src", "bootstrap")]
```
Comment (albertlarsan68) [inline x.py:25]: Why this change?

**Comment by albertlarsan68** [comment]
> > cache result of NixOS detection
> 
> Could you please do that in the Rust code too?

It was already done

File: x.py (hunk 3 above)
Comment (kadiwa4) [reply x.py:25]: Might make it a bit faster because it doesn't have to check all the other functions. Also theoretically someone could have something with the name `bootstrap` installed

Comment (albertlarsan68) [reply x.py:25]: So it is like a `prepend`, so that it can still find the other deps? (sorry, not python master)

Comment (kadiwa4) [reply x.py:25]: Yes, I'll add a short comment

File: src/bootstrap/download.rs (hunk 1 above)
Comment (kadiwa4) [reply src/bootstrap/download.rs:127]: In python there's currently no cache variable (`fix_bin_or_dylib` is only called once in that script), but I can add one

Comment (albertlarsan68) [reply src/bootstrap/download.rs:127]: I think having an assert in the Python side is also important, even if it takes the form of a global tuple `(init: Bool, value: Bool)` (pseudo syntax here). I think having the rust-side assert also asserting the init-ness of the cache, but is not necessary (after all, CI runs on Ubuntu, so it should not return true)

File: x.py (hunk 3 above)
Comment (albertlarsan68) [reply x.py:25]: Is doing it this way better than through `sys.path.insert(0, os.path.join(rust_dir, "src", "bootstrap"))`?
It seems clearer when you don't have a Python Indexing PhD

File: src/bootstrap/download.rs (hunk 1 above)
Comment (albertlarsan68) [reply src/bootstrap/download.rs:127]: Thanks!

File: x.py (hunk 3 above)
Comment (kadiwa4) [reply x.py:26]: I guess it isn't

File: src/bootstrap/bootstrap.py [hunk 4]
```diff
@@ -458,60 +460,73 @@ def _download_component_helper(
...
-        # At this point we're pretty sure the user is running NixOS or
-        # using Nix
-        nix_os_msg = "info: you seem to be using Nix. Attempting to patch"
-        print(nix_os_msg, fname)
+        This method is only required on NixOS and uses the PatchELF utility to
+        change the interpreter/RPATH of ELF executables.
+
+        Please see https://nixos.org/patchelf.html for more information
+        """
+        assert self._should_fix_bins_and_dylibs is True
```
[the remark this answers is not in the document]
Comment (albertlarsan68) [reply src/bootstrap/bootstrap.py:528]: What are the other possibilities to assert init-ness and True-ness?

[the remark this answers is not in the document]
Comment (kadiwa4) [reply src/bootstrap/bootstrap.py:528]: well, you can do `assert self._should_fix_bins_and_dylibs` and `None` will be cast to `False`

[the remark this answers is not in the document]
Comment (albertlarsan68) [reply src/bootstrap/bootstrap.py:528]: Does using `assert self._should_fix_bins_and_dylibs == True` works for this case? It is closer to the Rust code our fellow Rust contributor are knowing, so may be less confusing.

File: x.py (hunk 3 above)
Comment (albertlarsan68) [reply x.py]: You could have left your comment, it is always good to know why this path manipulation happens.

File: src/bootstrap/bootstrap.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kadiwa4) [reply src/bootstrap/bootstrap.py:528]: That works, but in Python `is` is considered idiomatic when checking for these constant objects

[the remark this answers is not in the document]
Comment (albertlarsan68) [reply src/bootstrap/bootstrap.py:528]: I think leaving it as `is` is good enough.

**Comment by albertlarsan68** [comment]
Thanks for your work!
@bors r+

File: src/bootstrap/bootstrap.py [hunk 5]
```diff
@@ -932,25 +921,32 @@ def main():
...
         else:
             exit_code = 1
             print(error)
-        if not help_triggered:
-            print("Build completed unsuccessfully in {}".format(
-                format_build_time(time() - start_time)))
-        sys.exit(exit_code)
+
+    if not help_triggered:
+        print("Build completed successfully in", format_build_time(time() - start_time))
```
Comment (jyn514) [inline src/bootstrap/bootstrap.py:948]: This is not correct. It no longer prints "Build completed unsuccessfully when bootstrap exits with an error.

Comment (kadiwa4) [reply src/bootstrap/bootstrap.py:948]: corrected by #107808
